File: backend/data/mock_data.py
"""
Python equivalent of lib/mockData.ts.

Generates the same hub-and-spoke campaign topology using a seeded RNG.
Node IDs are deterministic (prefix-bot-{cluster}-{index}) so the frontend's
sessionStorage-based campaign lookup continues to work after switching to
backend data.
"""

import random
from typing import List, Dict, Any


def _make_rand(prefix: str):
    seed = sum(ord(c) * (i + 7) for i, c in enumerate(prefix))
    rng = random.Random(seed)
    def rand(min_val: int, max_val: int) -> int:
        return rng.randint(min_val, max_val)
    return rand


def generate_network(
    prefix: str,
    cluster_count: int,
    bots_per_cluster: int,
    amplifier_count: int,
    legitimate_count: int = 0,
) -> Dict[str, List[Dict[str, Any]]]:
    rand = _make_rand(prefix)
    nodes: List[Dict[str, Any]] = []
    edges: List[Dict[str, Any]] = []

    # Origin (C2)
    origin_id = f"{prefix}-origin"
    nodes.append({
        "id":        origin_id,
        "type":      "origin",
        "label":     "C2 Origin",
        "accountId": "COVERT-C2",
        "posts":     0,
        "followers": 0,
    })

    all_bot_ids: List[str] = []

    for c in range(cluster_count):
        cluster_bots: List[str] = []
        for b in range(bots_per_cluster):
            bot_id = f"{prefix}-bot-{c}-{b}"
            all_bot_ids.append(bot_id)
            cluster_bots.append(bot_id)
            nodes.append({
                "id":        bot_id,
                "type":      "bot",
                "label":     f"BOT-C{c + 1}-{str(b + 1).zfill(2)}",
                "accountId": f"@usr_{rand(100000, 999999)}",
                "posts":     rand(200, 5000),
                "followers": rand(50, 2000),
                "clusterId": c,
            })
        # origin → cluster hub
        edges.append({"source": origin_id, "target": cluster_bots[0], "weight": 0.9})
        # hub → rest of cluster
        for b in range(1, len(cluster_bots)):
            edges.append({"source": cluster_bots[0], "target": cluster_bots[b], "weight": 0.6})

    all_amp_ids: List[str] = []
    for a in range(amplifier_count):
        amp_id = f"{prefix}-amp-{a}"
        all_amp_ids.append(amp_id)
        nodes.append({
            "id":        amp_id,
            "type":      "amplifier",
            "label":     f"AMP-{str(a + 1).zfill(3)}",
            "accountId": f"@acc_{rand(100000, 999999)}",
            "posts":     rand(50, 500),
            "followers": rand(200, 20000),
        })
        src = all_bot_ids[rand(0, len(all_bot_ids) - 1)]
        edges.append({"source": src, "target": amp_id, "weight": round(rand(25, 55) / 100, 2)})

    for lll in range(legitimate_count):
        leg_id = f"{prefix}-leg-{lll}"
        nodes.append({
            "id":        leg_id,
            "type":      "legitimate",
            "label":     f"USER-{rand(1000, 9999)}",
            "accountId": f"@real_{rand(100000, 999999)}",
            "posts":     rand(200, 2000),
            "followers": rand(500, 50000),
        })
        src = all_amp_ids[rand(0, len(all_amp_ids) - 1)]
        edges.append({"source": src, "target": leg_id, "weight": round(rand(10, 25) / 100, 2)})

    return {"nodes": nodes, "edges": edges}


def get_campaigns() -> List[Dict[str, Any]]:
    pulse      = generate_network("pulse",   4, 8, 20, 3)   # 56 nodes
    medfear    = generate_network("medfear", 3, 6, 15, 2)   # 36 nodes
    reviewstorm= generate_network("review",  2, 5, 10, 1)   # 22 nodes

    return [
        {
            "id":            "campaign-001",
            "name":          "Operation Pulse",
            "threat_level":  "HIGH",
            "account_count": 847,
            "start_time":    "2026-05-14T02:17:00Z",
            "narrative":     (
                "Coordinated election misinformation campaign targeting voter turnout in Maharashtra "
                "and Gujarat. False claims about EVM tampering and polling date changes amplified "
                "through bot networks seeded from three state-level coordination hubs."
            ),
            "confidence":    0.94,
            "nodes":         pulse["nodes"],
            "edges":         pulse["edges"],
        },
        {
            "id":            "campaign-002",
            "name":          "MedFear",
            "threat_level":  "HIGH",
            "account_count": 312,
            "start_time":    "2026-06-01T08:45:00Z",
            "narrative":     (
                "Anti-vaccine misinformation operation spreading fabricated WHO adverse-effect reports "
                "and deepfake audio attributed to medical professionals. Targets routine childhood "
                "vaccination programmes across tier-2 cities."
            ),
            "confidence":    0.91,
            "nodes":         medfear["nodes"],
            "edges":         medfear["edges"],
        },
        {
            "id":            "campaign-003",
            "name":          "ReviewStorm",
            "threat_level":  "MED",
            "account_count": 156,
            "start_time":    "2026-06-10T16:22:00Z",
            "narrative":     (
                "Coordinated fake product review network artificially inflating ratings for substandard "
                "consumer electronics while flooding competitor listings with negative sentiment. "
                "Active across Flipkart, Amazon IN, and Meesho."
            ),
            "confidence":    0.82,
            "nodes":         reviewstorm["nodes"],
            "edges":         reviewstorm["edges"],
        },
    ]

File: lib/mockData.ts
import type { Campaign, GraphNode, GraphEdge } from '@/types'

function makeRand(seed: number) {
  let s = (seed * 0x9e3779b9) | 1
  return (min: number, max: number): number => {
    s = Math.imul(s ^ (s >>> 16), 0x45d9f3b)
    s = Math.imul(s ^ (s >>> 16), 0x45d9f3b)
    s ^= s >>> 16
    const r = (s >>> 0) / 0x100000000
    return Math.floor(r * (max - min + 1)) + min
  }
}

function generateNetwork(
  prefix: string,
  clusterCount: number,
  botsPerCluster: number,
  amplifierCount: number,
  legitimateCount = 0,
): { nodes: GraphNode[]; edges: GraphEdge[] } {
  const seed = prefix.split('').reduce((a, c, i) => a + c.charCodeAt(0) * (i + 7), 0)
  const rand = makeRand(seed)

  const nodes: GraphNode[] = []
  const edges: GraphEdge[] = []

  // Origin node (C2 command-and-control)
  const originId = `${prefix}-origin`
  nodes.push({
    id: originId,
    type: 'origin',
    label: 'C2 Origin',
    accountId: 'COVERT-C2',
    posts: 0,
    followers: 0,
  })

  // Bot clusters — hub-and-spoke within each cluster
  const allBotIds: string[] = []

  for (let c = 0; c < clusterCount; c++) {
    const clusterBotIds: string[] = []

    for (let b = 0; b < botsPerCluster; b++) {
      const botId = `${prefix}-bot-${c}-${b}`
      allBotIds.push(botId)
      clusterBotIds.push(botId)
      nodes.push({
        id: botId,
        type: 'bot',
        label: `BOT-C${c + 1}-${String(b + 1).padStart(2, '0')}`,
        accountId: `@usr_${rand(100000, 999999)}`,
        posts: rand(200, 5000),
        followers: rand(50, 2000),
        clusterId: c,
      })
    }

    // origin → cluster hub (first bot)
    edges.push({ source: originId, target: clusterBotIds[0], weight: 0.9 })

    // hub → remaining bots in cluster
    for (let b = 1; b < clusterBotIds.length; b++) {
      edges.push({ source: clusterBotIds[0], target: clusterBotIds[b], weight: 0.6 })
    }
  }

  // Amplifiers — each connected to a random bot
  const allAmpIds: string[] = []
  for (let a = 0; a < amplifierCount; a++) {
    const ampId = `${prefix}-amp-${a}`
    allAmpIds.push(ampId)
    nodes.push({
      id: ampId,
      type: 'amplifier',
      label: `AMP-${String(a + 1).padStart(3, '0')}`,
      accountId: `@acc_${rand(100000, 999999)}`,
      posts: rand(50, 500),
      followers: rand(200, 20000),
    })
    edges.push({ source: allBotIds[rand(0, allBotIds.length - 1)], target: ampId, weight: 0.4 })
  }

  // Legitimate users — each connected to a random amplifier
  for (let l = 0; l < legitimateCount; l++) {
    const legId = `${prefix}-leg-${l}`
    nodes.push({
      id: legId,
      type: 'legitimate',
      label: `USER-${rand(1000, 9999)}`,
      accountId: `@real_${rand(100000, 999999)}`,
      posts: rand(200, 2000),
      followers: rand(500, 50000),
    })
    edges.push({ source: allAmpIds[rand(0, allAmpIds.length - 1)], target: legId, weight: 0.2 })
  }

  return { nodes, edges }
}

// Operation Pulse:  1 origin + 4×8 bots + 20 amps + 3 legitimate = 56 nodes
// MedFear:          1 origin + 3×6 bots + 15 amps + 2 legitimate = 36 nodes
// ReviewStorm:      1 origin + 2×5 bots + 10 amps + 1 legitimate = 22 nodes
const pulse       = generateNetwork('pulse',   4, 8, 20, 3)
const medfear     = generateNetwork('medfear', 3, 6, 15, 2)
const reviewstorm = generateNetwork('review',  2, 5, 10, 1)

export const campaigns: Campaign[] = [
  {
    id: 'campaign-001',
    name: 'Operation Pulse',
    threat_level: 'HIGH',
    account_count: 847,
    start_time: '2026-05-14T02:17:00Z',
    narrative:
      'Coordinated election misinformation campaign targeting voter turnout in Maharashtra and Gujarat. False claims about EVM tampering and polling date changes amplified through bot networks seeded from three state-level coordination hubs.',
    confidence: 0.94,
    nodes: pulse.nodes,
    edges: pulse.edges,
  },
  {
    id: 'campaign-002',
    name: 'MedFear',
    threat_level: 'HIGH',
    account_count: 312,
    start_time: '2026-06-01T08:45:00Z',
    narrative:
      'Anti-vaccine misinformation operation spreading fabricated WHO adverse-effect reports and deepfake audio attributed to medical professionals. Content targets routine childhood vaccination programmes across tier-2 cities.',
    confidence: 0.91,
    nodes: medfear.nodes,
    edges: medfear.edges,
  },
  {
    id: 'campaign-003',
    name: 'ReviewStorm',
    threat_level: 'MED',
    account_count: 156,
    start_time: '2026-06-10T16:22:00Z',
    narrative:
      'Coordinated fake product review network artificially inflating ratings for substandard consumer electronics while flooding competitor listings with negative sentiment. Operates across Flipkart, Amazon IN, and Meesho.',
    confidence: 0.82,
    nodes: reviewstorm.nodes,
    edges: reviewstorm.edges,
  },
]
